### DISCUSSION TITLE
Required GitHub token ... does not exist in private repo

### DISCUSSION THREAD

**Discussion by tillydray** [magit/forge#415] [Support] [answered]
I have read the forge manual and the ghub manual. I have read the issues in this repo as well as the support questions. None of them seem to be helping.

## Problem

Note that I have tried this on a public repo and I do not get this error. This must indicate an issue setting this up in the private repo, though I don't know what.

When I run `M-x forge-add-repository`, the repo url is already filled in and I hit ENTER. The next step I don't quite recall because it's been over an hour but I believe it asked for my username or email, and I think I entered <my_work_email> locally, not globally. Then I get the error `Required Github token ("<my_work_email>^forge" for "api.github.com") does not exist`.

## What I've done

- I have checked the permissions on the token and have even created a new token with _all_ scopes and I still get the error (related to https://github.com/magit/forge/issues/298).
- I have set my GitHub username in ~/.gitconfig (related to https://github.com/magit/forge/issues/201). The relevant parts of my gitconfig are below

I have also pasted below the debugging output when I do `M-x toggle-debug-on-error` and trigger the error mentioned above

<details>
  <summary>Relevant Git Configs</summary>
  
```
# ~/.gitconfig
[user]
[includeIf "gitdir:~/MotoRefi/"]
	path = ~/MotoRefi/.gitconfig
[gpg]
	program = gpg2
[github]
	user = JasonMFry
```
```
# ~/MotoRefi/.gitconfig
[user]
	email = <my_work_email>
```
  
</details>

<details>
  <summary>Debug Output</summary>

Note that I've replaced the repo name with <private_repo_name> 

```elisp
Debugger entered--Lisp error: (error "Required Github token (\"<my_work_email>^for...")
  signal(error ("Required Github token (\"<my_work_email>^for..."))
  error("Required %s token (%S for %S) does not exist.\nSee ..." "Github" "<my_work_email>^forge" "api.github.com")
  ghub--token("api.github.com" "<my_work_email>" forge nil nil)
  ghub--auth("api.github.com" forge "<my_work_email>" nil)
  #f(compiled-function () #<bytecode 0x1ff25b76a2f1>)()
  ghub--retrieve("{\"query\":\"query ($owner:String!, $name:String!) {\\..." #s(ghub--req :url #s(url :type "https" :user nil :password nil :host "api.github.com" :portspec nil :filename "/graphql" :target nil :attributes nil :fullness t :silent nil :use-cookies t :asynchronous t) :forge nil :silent nil :method "POST" :headers #f(compiled-function () #<bytecode 0x1ff25b76a2f1>) :handler ghub--handle-response :unpaginate nil :noerror nil :reader nil :callback nil :errorback nil :value nil :extra nil))
  ghub-request("POST" "/graphql" nil :payload "{\"query\":\"query ($owner:String!, $name:String!) {\\..." :headers nil :silent nil :username nil :auth forge :host "api.github.com" :callback nil :errorback nil :extra nil :value nil)
  ghub-graphql("query ($owner:String!, $name:String!) {\n          ..." ((owner . "motorefi") (name . "<private_repo_name>")) :username nil :auth forge :host "api.github.com")
  ghub--repository-id("motorefi" "<private_repo_name>" :username nil :auth forge :host "api.github.com")
  ghub-repository-id("motorefi" "<private_repo_name>" :host "api.github.com" :auth forge :forge github)
  #f(compiled-function (class host owner name &optional stub) "Return (OUR-ID . THEIR-ID) of the specified repository.\nIf optional STUB is non-nil, then the IDs are not guaranteed to\nbe unique.  Otherwise this method has to make an API request to\nretrieve THEIR-ID, the repository's ID on the forge.  In that\ncase OUR-ID derives from THEIR-ID and is unique across all\nforges and hosts." #<bytecode 0x1ff25bc33a95>)(forge-github-repository "github.com" "motorefi" "<private_repo_name>" nil)
  apply(#f(compiled-function (class host owner name &optional stub) "Return (OUR-ID . THEIR-ID) of the specified repository.\nIf optional STUB is non-nil, then the IDs are not guaranteed to\nbe unique.  Otherwise this method has to make an API request to\nretrieve THEIR-ID, the repository's ID on the forge.  In that\ncase OUR-ID derives from THEIR-ID and is unique across all\nforges and hosts." #<bytecode 0x1ff25bc33a95>) forge-github-repository ("github.com" "motorefi" "<private_repo_name>" nil))
  forge--repository-ids(forge-github-repository "github.com" "motorefi" "<private_repo_name>" nil)
  #f(compiled-function (&rest rest) "((host owner name) &optional remote demand)\n\nReturn the repository identified by HOST, OWNER and NAME." #<bytecode 0x1ff25bc3660d>)(("github.com" "motorefi" "<private_repo_name>") nil create)
  apply(#f(compiled-function (&rest rest) "((host owner name) &optional remote demand)\n\nReturn the repository identified by HOST, OWNER and NAME." #<bytecode 0x1ff25bc3660d>) ("github.com" "motorefi" "<private_repo_name>") (nil create))
  #f(compiled-function (arg &rest args) #<bytecode 0x1ff25757d1e9>)(("github.com" "motorefi" "<private_repo_name>") nil create)
  apply(#f(compiled-function (arg &rest args) #<bytecode 0x1ff25757d1e9>) (("github.com" "motorefi" "<private_repo_name>") nil create))
  forge-get-repository(("github.com" "motorefi" "<private_repo_name>") nil create)
  #f(compiled-function (url &optional remote demand) "Return the repository at URL." #<bytecode 0x1ff25bc35a91>)("git@github.com:motorefi/<private_repo_name>.git" nil create)
  apply(#f(compiled-function (url &optional remote demand) "Return the repository at URL." #<bytecode 0x1ff25bc35a91>) "git@github.com:motorefi/<private_repo_name>.git" (nil create))
  #f(compiled-function (arg &rest args) #<bytecode 0x1ff25757d1e9>)("git@github.com:motorefi/<private_repo_name>.git" nil create)
  apply(#f(compiled-function (arg &rest args) #<bytecode 0x1ff25757d1e9>) ("git@github.com:motorefi/<private_repo_name>.git" nil create))
  forge-get-repository("git@github.com:motorefi/<private_repo_name>.git" nil create)
  forge-add-repository("git@github.com:motorefi/<private_repo_name>.git")
  funcall-interactively(forge-add-repository "git@github.com:motorefi/<private_repo_name>.git")
  call-interactively(forge-add-repository record nil)
  command-execute(forge-add-repository record)
  counsel-M-x-action("forge-add-repository")
  ivy-call()
  ivy-read("M-x " ("toggle-debug-on-error" "forge-add-repository" "delete-file" "auth-source-forget-all-cached" "forge-forge.remote" "forge-reset-database" "forge-list-repositories" "package-install" "markdown-mode" "query-replace" "dired-do-touch" "magit-branch-create" "forge-remove-repository" "org-md-export-as-markdown" "cd" "5x5" "amx" "arp" "dbx" "dig" "erc" "ert" "eww" "ftp" "gdb" "irc" "jdb" "lsp" "man" "mpc" "pdb" "pwd" "rsh" "sdb" "sly" "xdb" "calc" "diff" "dirs" "ffap" "gnus" "grep" "help" "ielm" "info" "life" "lsp!" "mail" "mpuz" "mu4e" ...) :predicate counsel--M-x-externs-predicate :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  counsel-M-x()
  funcall-interactively(counsel-M-x)
  call-interactively(counsel-M-x nil nil)
  command-execute(counsel-M-x)
```

</details>

**Comment by tarsius** [comment]
> "<my_work_email>^forge"

Use your github username instead of `<my_work_email>`.

**Reply by tillydray** [reply]
Oops, that was something else I fixed before getting the debug output but forgot about. My `~/.authinfo.gpg` is as follows, and I am still getting the error.

`machine api.github.com login JasonMFry^forge password <token>`

**Reply by tillydray** [reply]
I think when I first tried to setup forge, when the prompt asked me for my username I used my work email address. This could be the source of the problem, but I don't know how to change it at this point. I have removed forge, restarted emacs, and added forge again in hopes that would restart the process but it didn't

**Reply by tarsius** [reply]
Try `git config --global github.user JasonMFry`.

**Reply by tillydray** [reply]
I did that too. My gitconfig contains
```
[github]
	user = JasonMFry
```

**Reply by tarsius** [reply]
You might have configure `auth-source` to not look in `~/.authinfo.gpg`. You might use an old buggy version of one of the involved packages. ...

**Reply by tillydray** [reply]
I'm using Doom Emacs, which already modifies `auth-sources` from the default. I further modified it to move `~/.authinfo.gpg` to the front of the list just in case (see below). I did that before posting here, but _just in case_ I just now modified it to be `(setq auth-sources '("~/.authinfo.gpg"))` but I'm still getting the same error.

```elisp
(setq auth-sources '("~/.authinfo.gpg" macos-keychain-generic macos-keychain-internet "~/.emacs.d/.local/etc/authinfo.gpg"))
```

**Reply by tillydray** [reply]
Is there a way to remove all forge config and start over, so I can get the prompt to add my username again? I tried that already by removing forge and restarting emacs, but that didn't do what I wanted.

**Reply by tillydray** [reply]
This is what I get when I remove forge and reload my config. Remember this is Doom Emacs. It seems to remove everything I would want, but after doing this and re-adding forge and doing `M-x forge-add-repository` I get the same experience as before with the same error.

```
    > Purging orphaned packages (for the emperor)...
      > Purging straight builds...
        ✓ Purged build/closql
        ✓ Purged build/emacsql
        ✓ Purged build/emacsql-sqlite
        ✓ Purged build/forge
        ✓ Purged build/orgit-forge
        ✓ Purged build/yaml
        Pruned forge from build cache
        Pruned closql from build cache
        Pruned emacsql-sqlite from build cache
        Pruned emacsql from build cache
        Pruned yaml from build cache
        Pruned orgit-forge from build cache
```

**Comment by tarsius** [comment, marked as answer]
> Is there a way to remove all forge config and start over, so I can get the prompt to add my username again?

```
git config --unset --local github.user
git config --unset --global github.user
```

You should then get the prompt:

```
Git variable `github.user' is unset.  Set to: 
```

and after answering that also:

```
Set github.user=JasonMFry [g]lobally (recommended) or [l]ocally? 
```

However, that won't do any good, because we already established that you are using the correct value for that.

If however you get a prompt like:

```
Username [for api.github.com]:
```

Then something is wrong.

**Reply by tillydray** [reply]
This helped me figure it out! `git config --unset --local github.user` did the trick, of course!

If I responded to the username prompt with <work_email> and told it local, then obviously the local config would have been modified, not the global one. What threw me off was I did all this at 6am after not sleeping well for 3 nights in a row, and that I kept looking at the config I created in `~/MotoRefi/.gitconfig` instead of the one local to the repo which is `~/MotoRefi/repo-name/.git/config`

**Reply by tarsius** [reply]
> What threw me off was I did all this at 6am after not sleeping well for 3 nights in a row,

I find such issues highly frustrating. Next time delay reporting an issue/asking for help until you had a chance to resolve it on your own with a clear mind.

**Comment by tarsius** [comment]
Have you seen https://github.com/magit/forge/wiki/Common-mistakes?

**Reply by tillydray** [reply]
Yes I read that. I'm not sure if my experience should be documented or not because it was somewhat specific and some silly mistakes on my part. But if you want me to I will add it to https://github.com/magit/forge/wiki/Common-mistakes or wherever you think is most appropriate.

**Reply by tarsius** [reply]
Yeah, please add a short note.

**Reply by tillydray** [reply]
@tarsius I just updated https://github.com/magit/forge/wiki/Common-mistakes, thanks again for your help!
